common/config/persistence.go

```diff
@@ -99,19 +109,21 @@ func (c *Persistence) Validate() error {
 		if err := ds.Validate(); err != nil {
 			return fmt.Errorf("persistence config: datastore %q: %s", st, err.Error())
 		}
-		if ds.Elasticsearch != nil {
-			cntEsConfigs++
-		}
 	}
-
-	if cntEsConfigs > 1 {
-		return fmt.Errorf(
-			"persistence config: cannot have more than one Elasticsearch visibility store config " +
-				"(use `elasticsearch.indices.secondary_visibility` config key if you need to set a " +
-				"secondary Elasticsearch visibility store)",
-		)
+	// ElasticSearch config for visibilityStore and secondaryVisibilityStore must be the same except for
+	// `indices.visibility` config key and private fields - this is a restriction due to global ES client
+	if c.DataStores[c.VisibilityStore].Elasticsearch != nil && c.DataStores[c.SecondaryVisibilityStore].Elasticsearch != nil {
+		visibilityStoreEsConfig := reflect.ValueOf(c.DataStores[c.VisibilityStore].Elasticsearch).Elem()
+		secondaryVisibilityStoreEsConfig := reflect.ValueOf(c.DataStores[c.SecondaryVisibilityStore].Elasticsearch).Elem()
+		for i := 0; i < max(visibilityStoreEsConfig.NumField(), secondaryVisibilityStoreEsConfig.NumField()); i++ {
+			if visibilityStoreEsConfig.Type().Field(i).Name == "Indices" || visibilityStoreEsConfig.Type().Field(i).PkgPath != "" {
+				continue
+			}
+			if !reflect.DeepEqual(visibilityStoreEsConfig.Field(i).Interface(), secondaryVisibilityStoreEsConfig.Field(i).Interface()) {
+				return fmt.Errorf("%w: %q", ErrConfigMismatch, visibilityStoreEsConfig.Type().Field(i).Name)
+			}
+		}
```

Comment:
This looks unnecessarily complex. Wouldn't this work?

```suggestion
		esConfig := *c.DataStores[c.VisibilityStore].Elasticsearch
		secEsConfig := *c.DataStores[c.SecondaryVisibilityStore].Elasticsearch
		esConfig.Indices = nil
		secEsConfig.Indices = nil
		if !reflect.DeepEqual(esConfig, secEsConfig) {
			return ErrConfigMismatch
		}
```

Reply:
yes this should also work, I updated and tested the code. Thanks for the suggestion!